src/sentry/sentry_metrics/indexer/base.py

```diff
@@ -1,8 +1,162 @@
-from typing import Mapping, Optional, Set
+from collections import defaultdict
+from dataclasses import dataclass
+from enum import Enum
+from typing import (
+    Mapping,
+    MutableMapping,
+    MutableSequence,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+)
 
 from sentry.utils.services import Service
 
 
+class FetchType(Enum):
+    CACHE_HIT = "c"
+    HARDCODED = "h"
+    DB_READ = "d"
+    FIRST_SEEN = "f"
+
+
+KR = TypeVar("KR", bound="KeyResult")
+KRs = TypeVar("KRs", bound="KeyResults")
+
+
+@dataclass(frozen=True)
+class KeyResult:
+    org_id: int
+    string: str
+    id: int
+
+    @classmethod
+    def from_string(cls: Type[KR], key: str, id: int) -> KR:
+        org_id, string = key.split(":", 1)
+        return cls(int(org_id), string, id)
+
+
+class KeyCollection:
+    """
+    A KeyCollection is a way of keeping track of a group of keys
+    used to fetch ids, whose results are stored in KeyResults.
+
+    A key is a org_id, string pair, either represented as a
+    tuple e.g (1, "a"), or a string "1:a".
+
+    Initial mapping is org_id's to sets of strings:
+        { 1: {"a", "b", "c"}, 2: {"e", "f"} }
+    """
+
+    def __init__(self, mapping: Mapping[int, Set[str]]):
+        self.mapping = mapping
+        self.size = self._size()
+
+    def _size(self) -> int:
+        total_size = 0
+        for org_id in self.mapping.keys():
+            total_size += len(self.mapping[org_id])
+        return total_size
+
+    def as_tuples(self) -> Sequence[Tuple[int, str]]:
+        """
+        Returns all the keys, each key represented as tuple -> (1, "a")
+        """
+        key_pairs: MutableSequence[Tuple[int, str]] = []
+        for org_id in self.mapping:
+            key_pairs.extend([(org_id, string) for string in self.mapping[org_id]])
+
+        return key_pairs
+
+    def as_strings(self) -> Sequence[str]:
+        """
+        Returns all the keys, each key represented as string -> "1:a"
+        """
+        keys: MutableSequence[str] = []
+        for org_id in self.mapping:
+            keys.extend([f"{org_id}:{string}" for string in self.mapping[org_id]])
+
+        return keys
+
+
+class KeyResults:
```

Comment:
Most of this class is just copied from `postgres_v2.py` so that we could change the method signature of bulk_record on the base class